### ISSUE TITLE
component implements ControlValueAccessor still connected to ngForm after destroy

### ISSUE THREAD

**Issue by ydomenjoud** [angular/angular#27803]
# 🐞 bug report

### Affected Package
The issue is caused by package @angular/forms

### Is this a regression?
I don't know if this had work before.

### Description
When a component implementing ControlValueAccessor is destroyed, writeValue is still called on patchValue for example. Destroy component multiple times results in multiple writeValue called

## 🔬 Minimal Reproduction
please try this stackblitz, click on [toggle input] button multiple times to create and destroy component and click on patch value button to see in console the console.log from writeValue
https://stackblitz.com/edit/angular-pfmvxh
```
 writeValue(obj: any): void {
    console.log('writeValue', obj);
  }
```
You should see one writeValue by component destroyed

## 🔥 Exception or Error
if detection strategyis set  to OnPush and call ChangeDetectorRef.detectChanges() in writeValue we will raise a ViewDestroyedError that confirm that writeValue is called even component is destroyed

## 🌍  Your Environment

**Angular Version:**
<pre><code>
@angular-devkit/architect         0.10.6
@angular-devkit/build-angular     0.10.6
@angular-devkit/build-optimizer   0.10.6
@angular-devkit/build-webpack     0.10.6
@angular-devkit/core              7.0.6
@angular-devkit/schematics        7.0.6
@angular/cdk                      7.1.0
@angular/cli                      7.0.6
@angular/material                 7.1.1
</code></pre>

**Comment by maxime1992** [comment]
I was going nut. For the last 30mn I've been trying to understand what was wrong in code :sweat_smile:.

A bit different but on the same principle: It's also calling `validate` when implementing `Validator`.

That is really annoying, not sure how I'm going to solve that in the meantime :thinking:...

In my case I have a toggle and based on that I display either one component or another.

The first one doesn't have any validation required so it's all fine but when I toggle and go on the second one, I get an error. Which is fine as something's required and not filled by default. BUT. When I toggle once again, it's still invalid.

**Comment by Airblader** [comment]
I feel like it's pretty clear what's wrong. When the component is created (and with it the formControl directive), it calls [`FormGroupDirective#addControl`](https://github.com/angular/angular/blob/master/packages/forms/src/directives/reactive_directives/form_group_directive.ts#L130):

```
  addControl(dir: FormControlName): FormControl {
    const ctrl: any = this.form.get(dir.path);
    setUpControl(ctrl, dir);
    ctrl.updateValueAndValidity({emitEvent: false});
    this.directives.push(dir);
    return ctrl;
  }
```

The `setUpControl` method here wires everything up so that `writeValue` is called etc. However, the FormControlDirective doesn't have the code that FormNameDirective has:

```
ngOnDestroy(): void {
    if (this.formDirective) {
      this.formDirective.removeControl(this);
    }
  }
```

We probably just need to add this.

**Comment by maxime1992** [comment]
I've found (an abandoned) PR https://github.com/angular/angular/pull/22503

There's also few depulicates:
https://github.com/angular/angular/issues/21434
https://github.com/angular/angular/issues/26890
https://github.com/angular/angular/issues/18414

@Airblader glad that `it's pretty clear what's wrong`. A fix would be more than welcome :raised_hands:.

Might be also worth taking a look into the opened PR :)

**Comment by Airblader** [comment]
> A fix would be more than welcome

I would, I just don't have time right now :-) I just recently went through this code, so I figured I could throw in what I expect the reason to be. The abandoned PR seems to confirm that.

**Comment by longgt** [comment]
Duplicated with https://github.com/angular/angular/issues/20007

**Comment by maxime1992** [comment]
Commenting here before the bot closes that issue too... As the bug is still here :bug:

**Comment by danielwegener** [comment]
🐛 🐛

**Comment by Airblader** [comment]
Guys, the bot only locks already closed issues, it does not close open issues due to inactivity. There's a bunch of people receiving emails for every comment so I would appreciate if we could keep the comments relevant to the issue and not succumb to spamming it with emojis.

**Comment by MuhammadElhrz** [comment]
Is there any fix or workaround for that bug?

**Comment by biancaaaaa** [comment]
I found a workaround for this issue, which is basically just removing the control from the form group and adding it again:
```
this.formGroup.removeControl("condition");
this.formGroup.addControl("condition", new FormControl(null, Validators.required));
```
Not very pretty, but at least it solves the issue of `writeValue` and `validate` still being called.

**Comment by dev054** [comment]
Wasn't it fixed by #39623?

cc @AndrewKushnir.

**Comment by AndrewKushnir** [comment]
@dev054 yes, this issue should be resolved by #39235 (which relied on the changes from #39623 that you mentioned). The changes from #39235 would be available in v11.1.0, please create a new ticket if the problem still exists after upgrading to v11.1.0 (once it's available on NPM later this month). Thank you.
